### DISCUSSION TITLE
Non quadratic constraints are not allowed

### DISCUSSION THREAD

**Discussion by readygo67** [SutuLabs/zkcourse#7] [Week2] [answered]
在下面的代码中，如果main= Max0(), 或者main =Max1(), 电路都能正常运行，在main=Max2()的时候，zkrepl提示Non quadratic constraints are not allowed!， 大家知道具体原因吗？

```
pragma circom 2.1.4;

function fmax(a,b){
    var max = a>b ?a:b;
    return max;
}

template Max0(){
    signal input in[2];
    signal output out;
    signal max;

    max <-- in[0] > in[1] ? in[0]: in[1];
    out <== max;
}

template Max1(){
    signal input in[2];
    signal output out;
    signal max;

    max <-- fmax(in[0], in[1]);
    out <== max;
}

template Max2(){
    signal input in[2];
    signal output out;
    var max;

    max = fmax(in[0], in[1]);
    out <== max;
}

component main = Max2();  //Max0(), Max1() works well 

/* INPUT = {
    "in": ["8", "5"]
} */
```

<img width="1212" alt="image" src="https://user-images.githubusercontent.com/78890754/233989514-5ce021aa-65f2-4d6b-8a03-6c48f9a8a702.png">

**Comment by readygo67** [comment]
类似的， 
```circom
pragma circom 2.1.4;


template AND(){
    signal input a;
    signal input b;
    signal output out;

    out <== a&b;
//     out[1] <== a|b;
//     out[2] <== ~a;
//     out[3] <== a^b;
}


component main = AND();

/* INPUT = {
    "a": "7", //0x111
    "b": "8   //0x1000
} */
```
<img width="1217" alt="image" src="https://user-images.githubusercontent.com/78890754/233997548-c8e74654-d7f3-44cf-9b7d-0a3317c36bc6.png">

**Comment by wizicer** [comment]
Max0和Max1之所以没有报错，是因为没有约束产生，可以注意看一下输出。

Max2出现的错误是试图产生约束时，发现约束无法被正确的写出。

&(AND)不是未知信号(unknown signal)可以使用的操作符，需要使用上课时讲解的数学表达式方法进行编写。

**Reply by readygo67** [reply]
@wizicer  对于“Max0和Max1之所以没有报错，是因为没有约束产生，可以注意看一下输出。”这句话，我的理解如下，你看是否正确？
1. Max0/1 中因为没有对max 做约束，不会报错。那是不是意味着proof 无法对max提供证明，进而整个证明的链条是有漏洞的？
2. 输出如下，如果1成立，那么输出中的“Everything went okay, circom safe”  是不是会误导？ <img width="1220" alt="image" src="https://user-images.githubusercontent.com/78890754/234052890-733cb9ca-631d-4113-9e3f-fb9998eb6c33.png">

**Comment by wenjin1997** [comment, marked as answer]
可以参考Circom文档[Constraint Generation](https://docs.circom.io/circom-language/constraint-generation/)，所有约束必须是二次表达式。

**Reply by readygo67** [reply]
这个文档讲的很清楚了，感谢！
